src/main/java/io/javalin/core/compression/DynamicCompressionHandler.kt

```diff
@@ -0,0 +1,63 @@
+package io.javalin.core.compression
+
+import io.javalin.core.JavalinConfig
+import io.javalin.core.util.Header
+import io.javalin.http.Context
+import javax.servlet.http.HttpServletResponse
+
+/**
+ * Class designed to handle dynamic content compression for Javalin.
+ *
+ * Uses DynamicCompressionStrategy within JavalinConfig to determine the type and level of compression to be applied.
+ *
+ * @see DynamicCompressionStrategy
+ */
+class DynamicCompressionHandler(val ctx: Context, val config: JavalinConfig) {
+
+    val compressionStrategy = config.inner.dynamicCompressionStrategy
+
+    /**
+     * @param res The response that we wish to encode
+     */
+    fun compressResponse(res: HttpServletResponse) {
+        val resultStream = ctx.resultStream()!!
+        if (brotliShouldBeDone(ctx)) { //Do Brotli
+            val level = compressionStrategy?.brotliLevel ?: DynamicCompressionStrategy.BROTLI_DEFAULT_LEVEL
+            BrotliWrapper(level).use { brWrappwer ->
+                res.setHeader(Header.CONTENT_ENCODING, "br")
+                res.outputStream.write(brWrappwer.compressByteArray(resultStream.readBytes()))
+            }
+            return
+        } else if (gzipShouldBeDone(ctx)) { //Do GZIP
+            val level = compressionStrategy?.gzipLevel ?: DynamicCompressionStrategy.GZIP_DEFAULT_LEVEL
+            GZIPWrapper(res.outputStream, true).setLevel(level).use { gzippedStream ->
```

Comment:
It seems wrong to set the level every time you use it. I imagine the app shouldn't change compression level between requests.

Reply:
I set the level on every call simply because a new GzipWrapper object is created every time that line of code is run. The underlying GZIPOutputStream that this class extends was being used exactly the same way, it just didn't allow for level setting.

I could move level into the constructor as a parameter , if preferred? This would also bring it in line with the BrotliWrapper object creation. Is that what you meant?

Reply:
The GzipWrapper class has been converted to Kotlin, and passing the level setting is now done via constructor.